src/test/spec/unified_runner/operation.rs

```diff
@@ -93,16 +103,104 @@ pub trait TestOperation: Debug + Send + Sync {
     }
 }
 
+macro_rules! with_mut_session {
+    ($test_runner:ident, $id:expr, |$session:ident| $body:expr) => {
+        async {
+            let id = $id;
+            let mut session_owned = match $test_runner.entities.write().await.remove(id).unwrap() {
+                Entity::Session(session_owned) => session_owned,
+                o => panic!(
+                    "expected {} to be a session entity, instead was {:?}",
+                    $id, o
+                ),
+            };
+            let $session = &mut session_owned;
+            let out = $body.await;
+            $test_runner
+                .entities
+                .write()
+                .await
+                .insert(id.to_string(), Entity::Session(session_owned));
+            out
+        }
+    };
+}
+
 #[derive(Debug)]
-pub struct Operation {
+pub(crate) struct Operation {
     operation: Box<dyn TestOperation>,
-    pub name: String,
-    pub object: OperationObject,
-    pub expectation: Expectation,
+    pub(crate) name: String,
+    pub(crate) object: OperationObject,
+    pub(crate) expectation: Expectation,
+}
+
+impl Operation {
+    pub(crate) async fn execute<'a>(&self, test_runner: TestRunner, description: &str) {
```

Comment:
this was just copy/pasted here so that we could call `operation.execute` from test runner threads. Previously this was all inline in the runner.